### DISCUSSION TITLE
How can I create a mysql trigger using sqlalchemy?

### DISCUSSION THREAD

**Discussion by BlackLotus** [sqlalchemy/sqlalchemy#11898] [Usage Questions] [answered]
I'm really confused I got this trigger that I can execute in a mysql shell that then just works
```
DELIMITER $$

CREATE TRIGGER before_insert_projects
BEFORE INSERT ON projects
FOR EACH ROW
BEGIN
  DECLARE max_counter INT;

  -- Find the maximum counter for the owner_id
  SELECT IFNULL(MAX(counter), -1) INTO max_counter
  FROM projects
  WHERE owner_id = NEW.owner_id;

  -- Set the new counter value to max_counter + 1
  SET NEW.counter = max_counter + 1;
END$$

DELIMITER ;
```
But no matter what I try I can't seem to add it using sqlalchemy only. Connection executes don't work and every sort of abstraction I tried failed in some way or another...
How can I add this trigger using sqlalchemy?

**Comment by zzzeek** [comment]
depends on the driver you're using, likely the "DELIMITER" stuff is not interpreted by the driver as expected.      the specific nature of "failed" is always very important to being able to answer these questions also.

**Reply by BlackLotus** [reply]
It's a mariadb and I tried using `mysql+pymysql` I tried so many possible code combinations that I felt stupid posting every single one. I can try to reconstruct some of my tries and post the error messages

**Reply by zzzeek** [reply, marked as answer]
just remove the DELIMETER statements, they don't apply here

run this

```py
from sqlalchemy import create_engine


e = create_engine("mysql+pymysql://scott:tiger@localhost/test", echo=True)

with e.begin() as conn:
    conn.exec_driver_sql("DROP TABLE IF EXISTS projects")
    conn.exec_driver_sql("""
CREATE TABLE projects (
    owner_id INT,
    counter INT
)
""")

    conn.exec_driver_sql("""


CREATE TRIGGER before_insert_projects
BEFORE INSERT ON projects
FOR EACH ROW
BEGIN
  DECLARE max_counter INT;

  -- Find the maximum counter for the owner_id
  SELECT IFNULL(MAX(counter), -1) INTO max_counter
  FROM projects
  WHERE owner_id = NEW.owner_id;

  -- Set the new counter value to max_counter + 1
  SET NEW.counter = max_counter + 1;
END


""")


```

**Reply by BlackLotus** [reply]
I swear I tested a nearly identical piece of code and it didn't work, but I probably removed too much/too little... anyway thank you VERY much. I was going crazy ^^
